File: server/legacy_js/routes/assignments.js
const express = require('express');
const router = express.Router();
const { query } = require('../db/connection');
const { authorizeRole } = require('../middleware/auth');

// Get all assignments
router.get('/', authorizeRole(['admin', 'teacher', 'student', 'parent', 'super_admin']), async (req, res) => {
  try {
    const { user } = req;
    let sql = 'SELECT a.*, c.name as course_name, c.grade FROM assignments a JOIN courses c ON a.course_id = c.id WHERE c.school_id = $1';
    const params = [req.schoolId];

    if (user.role === 'teacher') {
      sql += ' AND c.teacher_id = $2';
      params.push(user.id);
    } else if (user.role === 'parent') {
      // For parents, get assignments for courses their children are enrolled in
      sql += ` AND c.grade IN (
        SELECT DISTINCT s.grade
        FROM students s
        JOIN parent_student_relations psr ON s.id = psr.student_id
        WHERE psr.parent_id = $2 AND s.school_id = $1 AND s.status = 'active'
      )`;
      params.push(user.id);
    }

    sql += ' ORDER BY a.due_date DESC';
    const result = await query(sql, params);
    res.json({ success: true, data: result.rows });
  } catch (err) {
    console.error('Error fetching assignments:', err);
    res.status(500).json({ success: false, error: 'Failed to fetch assignments' });
  }
});

// Get assignment by ID
router.get('/:id', authorizeRole(['admin', 'teacher', 'student']), async (req, res) => {
  try {
    const { id } = req.params;

    const result = await query(
      'SELECT a.*, c.name as course_name FROM assignments a JOIN courses c ON a.course_id = c.id WHERE a.id = $1 AND c.school_id = $2',
      [id, req.schoolId]
    );
    
    if (result.rows.length === 0) {
      return res.status(404).json({ success: false, error: 'Assignment not found' });
    }
    
    res.json({ success: true, data: result.rows[0] });
  } catch (err) {
    console.error('Error fetching assignment:', err);
    res.status(500).json({ success: false, error: 'Failed to fetch assignment' });
  }
});

// Create assignment
router.post('/', authorizeRole(['admin', 'teacher']), async (req, res) => {
  try {
    const { schoolId, user } = req;
    const { course_id, title, description, due_date, total_marks } = req.body;

    // For teachers, validate they own the course
    if (user.role === 'teacher') {
      const courseCheck = await query(
        'SELECT id FROM courses WHERE id = $1 AND teacher_id = $2 AND school_id = $3',
        [course_id, user.id, schoolId]
      );

      if (courseCheck.rows.length === 0) {
        return res.status(403).json({ success: false, error: 'You can only create assignments for courses you teach' });
      }
    }

    const result = await query(
      'INSERT INTO assignments (school_id, course_id, teacher_id, title, description, due_date, total_marks) VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *',
      [schoolId, course_id, user.id, title, description, due_date, total_marks]
    );

    res.status(201).json({ success: true, data: result.rows[0], message: 'Assignment created successfully' });
  } catch (err) {
    console.error('Error creating assignment:', err);
    res.status(500).json({ success: false, error: 'Failed to create assignment' });
  }
});

// Update assignment
router.put('/:id', authorizeRole(['admin', 'teacher']), async (req, res) => {
  try {
    const { user } = req;
    const { id } = req.params;
    const { title, description, due_date, total_marks } = req.body;

    // For teachers, validate they own the course this assignment belongs to
    if (user.role === 'teacher') {
      const ownershipCheck = await query(
        'SELECT a.id FROM assignments a JOIN courses c ON a.course_id = c.id WHERE a.id = $1 AND c.teacher_id = $2 AND c.school_id = $3',
        [id, user.id, req.schoolId]
      );

      if (ownershipCheck.rows.length === 0) {
        return res.status(403).json({ success: false, error: 'You can only update assignments for courses you teach' });
      }
    }

    const result = await query(
      `UPDATE assignments SET title = $1, description = $2, due_date = $3, total_marks = $4
       WHERE id = $5 RETURNING *`,
      [title, description, due_date, total_marks, id]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ success: false, error: 'Assignment not found' });
    }

    res.json({ success: true, data: result.rows[0], message: 'Assignment updated successfully' });
  } catch (err) {
    console.error('Error updating assignment:', err);
    res.status(500).json({ success: false, error: 'Failed to update assignment' });
  }
});

// Delete assignment
router.delete('/:id', authorizeRole(['admin', 'teacher']), async (req, res) => {
  try {
    const { user } = req;
    const { id } = req.params;

    // For teachers, validate they own the course this assignment belongs to
    if (user.role === 'teacher') {
      const ownershipCheck = await query(
        'SELECT a.id FROM assignments a JOIN courses c ON a.course_id = c.id WHERE a.id = $1 AND c.teacher_id = $2 AND c.school_id = $3',
        [id, user.id, req.schoolId]
      );

      if (ownershipCheck.rows.length === 0) {
        return res.status(403).json({ success: false, error: 'You can only delete assignments for courses you teach' });
      }
    }

    await query('DELETE FROM assignments WHERE id = $1', [id]);
    res.json({ success: true, message: 'Assignment deleted successfully' });
  } catch (err) {
    console.error('Error deleting assignment:', err);
    res.status(500).json({ success: false, error: 'Failed to delete assignment' });
  }
});

// Get submissions for an assignment
router.get('/:id/submissions', authorizeRole(['admin', 'teacher']), async (req, res) => {
  try {
    const { id } = req.params;

    const result = await query(
      `SELECT asub.id, asub.assignment_id, asub.student_id, asub.submission_text, asub.attachment_url,
              asub.submitted_at, asub.grade as score, asub.feedback, asub.status,
              s.first_name, s.last_name, s.admission_number
       FROM assignment_submissions asub
       JOIN students s ON asub.student_id = s.id
       JOIN assignments a ON asub.assignment_id = a.id
       WHERE asub.assignment_id = $1 AND a.school_id = $2
       ORDER BY asub.submitted_at DESC`,
      [id, req.schoolId]
    );

    res.json({ success: true, data: result.rows });
  } catch (err) {
    console.error('Error fetching submissions:', err);
    res.status(500).json({ success: false, error: 'Failed to fetch submissions' });
  }
});

// Submit assignment (student)
router.post('/:id/submit', authorizeRole(['student']), async (req, res) => {
  try {
    const { user } = req;
    const { id } = req.params;
    const { content, attachment_url } = req.body;
    
    // Get student ID
    const studentResult = await query('SELECT id FROM students WHERE school_id = $1 AND email = $2', [req.schoolId, user.email]);
    
    if (studentResult.rows.length === 0) {
      return res.status(404).json({ success: false, error: 'Student not found' });
    }
    
    const studentId = studentResult.rows[0].id;
    
    const result = await query(
      `INSERT INTO assignment_submissions (assignment_id, student_id, content, attachment_url, status, submitted_at)
       VALUES ($1, $2, $3, $4, 'submitted', NOW())
       ON CONFLICT (assignment_id, student_id) 
       DO UPDATE SET content = $3, attachment_url = $4, status = 'submitted', submitted_at = NOW()
       RETURNING *`,
      [id, studentId, content, attachment_url]
    );
    
    res.status(201).json({ success: true, data: result.rows[0], message: 'Assignment submitted successfully' });
  } catch (err) {
    console.error('Error submitting assignment:', err);
    res.status(500).json({ success: false, error: 'Failed to submit assignment' });
  }
});

// Grade submission
router.post('/submissions/:submissionId/grade', authorizeRole(['admin', 'teacher']), async (req, res) => {
  try {
    const { user } = req;
    const { submissionId } = req.params;
    const { grade, feedback } = req.body;

    // For teachers, validate they own the course this submission belongs to
    if (user.role === 'teacher') {
      const ownershipCheck = await query(
        `SELECT asub.id FROM assignment_submissions asub
         JOIN assignments a ON asub.assignment_id = a.id
         JOIN courses c ON a.course_id = c.id
         WHERE asub.id = $1 AND c.teacher_id = $2 AND c.school_id = $3`,
        [submissionId, user.id, req.schoolId]
      );

      if (ownershipCheck.rows.length === 0) {
        return res.status(403).json({ success: false, error: 'You can only grade submissions for courses you teach' });
      }
    }

    const result = await query(
      `UPDATE assignment_submissions SET grade = $1, feedback = $2, status = 'graded', graded_by = $3, graded_at = NOW(), updated_at = NOW()
       WHERE id = $4 RETURNING *`,
      [grade, feedback, user.id, submissionId]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ success: false, error: 'Submission not found' });
    }

    res.json({ success: true, data: result.rows[0], message: 'Submission graded successfully' });
  } catch (err) {
    console.error('Error grading submission:', err);
    res.status(500).json({ success: false, error: 'Failed to grade submission' });
  }
});

// Download assignment responses
router.get('/:id/download', authorizeRole(['admin', 'teacher']), async (req, res) => {
  try {
    const { user } = req;
    const { id } = req.params;

    // For teachers, validate they own the course this assignment belongs to
    if (user.role === 'teacher') {
      const ownershipCheck = await query(
        'SELECT a.id FROM assignments a JOIN courses c ON a.course_id = c.id WHERE a.id = $1 AND c.teacher_id = $2 AND c.school_id = $3',
        [id, user.id, req.schoolId]
      );

      if (ownershipCheck.rows.length === 0) {
        return res.status(403).json({ success: false, error: 'You can only download responses for courses you teach' });
      }
    }

    // Get assignment details
    const assignmentResult = await query('SELECT title FROM assignments WHERE id = $1 AND school_id = $2', [id, req.schoolId]);
    if (assignmentResult.rows.length === 0) {
      return res.status(404).json({ success: false, error: 'Assignment not found' });
    }

    // Get all submissions
    const submissionsResult = await query(
      `SELECT s.first_name, s.last_name, s.admission_number, asub.submission_text, asub.attachment_url,
              asub.grade, asub.feedback, asub.status, asub.submitted_at
       FROM assignment_submissions asub
       JOIN students s ON asub.student_id = s.id
       WHERE asub.assignment_id = $1
       ORDER BY asub.submitted_at DESC`,
      [id]
    );

    // Generate CSV content
    const csvHeader = 'Student Name,Admission Number,Submission,Attachment URL,Grade,Feedback,Status,Submitted At\n';
    const csvRows = submissionsResult.rows.map(sub => {
      const studentName = `"${sub.first_name} ${sub.last_name}"`;
      const admissionNumber = `"${sub.admission_number || ''}"`;
      const submission = `"${(sub.submission_text || '').replace(/"/g, '""')}"`;
      const attachmentUrl = `"${sub.attachment_url || ''}"`;
      const grade = `"${sub.grade || ''}"`;
      const feedback = `"${(sub.feedback || '').replace(/"/g, '""')}"`;
      const status = `"${sub.status}"`;
      const submittedAt = `"${new Date(sub.submitted_at).toLocaleString()}"`;

      return `${studentName},${admissionNumber},${submission},${attachmentUrl},${grade},${feedback},${status},${submittedAt}`;
    });

    const csvContent = csvHeader + csvRows.join('\n');

    // Set headers for file download
    const filename = `assignment_${id}_responses_${new Date().toISOString().split('T')[0]}.csv`;
    res.setHeader('Content-Type', 'text/csv');
    res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);

    res.send(csvContent);
  } catch (err) {
    console.error('Error downloading responses:', err);
    res.status(500).json({ success: false, error: 'Failed to download responses' });
  }
});

module.exports = router;

File: server/assignments.py
# server/assignments.py
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from auth import get_current_user
from database import get_db
from models import Assignment, AssignmentSubmission, Course, Student, StudentCourseEnrollment, User

router = APIRouter(prefix="/api/assignments", tags=["Assignments"])


class GradeSubmissionRequest(BaseModel):
    grade: float
    feedback: Optional[str] = None


class AssignmentCreate(BaseModel):
    title: str
    course_id: int
    due_date: str          # ISO date or datetime string, e.g. "2026-07-20T14:30"
    max_score: Optional[int] = 100
    assignment_type: Optional[str] = "homework"
    description: Optional[str] = None
    instructions: Optional[str] = None


@router.post("")
@router.post("/")
async def create_assignment(
    payload: AssignmentCreate,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """Create a new assignment for the current teacher's school, and seed a
    pending submission row for every student enrolled in the course so the
    teacher's grading view shows the full class roster from day one."""
    user, token_payload = current_user
    school_id = token_payload.get("school_id")

    if not school_id:
        raise HTTPException(status_code=403, detail="Access token is not scoped to a specific school")

    course_res = await db.execute(
        select(Course).where(Course.id == payload.course_id, Course.school_id == school_id)
    )
    course = course_res.scalar_one_or_none()
    if not course:
        raise HTTPException(status_code=404, detail="Selected subject/course was not found.")

    try:
        due_date = datetime.fromisoformat(payload.due_date)
    except ValueError:
        raise HTTPException(status_code=422, detail="due_date must be a valid ISO date or datetime string")

    assignment = Assignment(
        school_id=school_id,
        course_id=payload.course_id,
        teacher_id=user.id,
        title=payload.title,
        description=payload.description,
        instructions=payload.instructions,
        assignment_type=payload.assignment_type or "homework",
        total_marks=payload.max_score or 100,
        due_date=due_date,
    )
    db.add(assignment)
    await db.flush()  # get assignment.id without a second round trip

    enrolled_res = await db.execute(
        select(StudentCourseEnrollment.student_id).where(
            StudentCourseEnrollment.course_id == payload.course_id,
            StudentCourseEnrollment.school_id == school_id,
        )
    )
    for (student_id,) in enrolled_res.all():
        db.add(AssignmentSubmission(
            assignment_id=assignment.id,
            student_id=student_id,
            status="pending",
        ))

    await db.commit()
    await db.refresh(assignment)

    return {
        "data": {
            "id": assignment.id,
            "title": assignment.title,
            "course_id": assignment.course_id,
            "course_name": course.name,
            "due_date": assignment.due_date.isoformat() if assignment.due_date else None,
            "status": assignment.status,
            "total_marks": assignment.total_marks,
        }
    }


@router.get("")
@router.get("/")
async def get_all_assignments(
    course_id: Optional[int] = None,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """Fetch this teacher's own assignments for their current school."""
    user, payload = current_user
    school_id = payload.get("school_id")

    if not school_id:
        return {"data": []}

    query = (
        select(Assignment, Course.name)
        .join(Course, Course.id == Assignment.course_id)
        .where(Assignment.school_id == school_id, Assignment.teacher_id == user.id)
    )
    if course_id is not None:
        query = query.where(Assignment.course_id == course_id)
    query = query.order_by(Assignment.due_date.desc().nullslast(), Assignment.created_at.desc())

    result = await db.execute(query)
    rows = result.all()

    data = []
    for assign, course_name in rows:
        submission_counts = await db.execute(
            select(AssignmentSubmission.status).where(AssignmentSubmission.assignment_id == assign.id)
        )
        statuses = [s for (s,) in submission_counts.all()]
        data.append({
            "id": assign.id,
            "title": assign.title,
            "course_id": assign.course_id,
            "course_name": course_name,
            "due_date": assign.due_date.isoformat() if assign.due_date else None,
            "status": assign.status,
            "assignment_type": assign.assignment_type,
            "total_marks": assign.total_marks,
            "total_students": len(statuses),
            "graded_count": sum(1 for s in statuses if s == "graded"),
            "submitted_count": sum(1 for s in statuses if s in ("submitted", "graded")),
        })

    return {"data": data}


@router.get("/{assignment_id}")
async def get_assignment_detail(assignment_id: int, db: AsyncSession = Depends(get_db), current_user=Depends(get_current_user)):
    """Fetch assignment details"""
    user, payload = current_user
    school_id = payload.get("school_id")

    result = await db.execute(
        select(Assignment, Course.name)
        .join(Course, Course.id == Assignment.course_id)
        .where(
            Assignment.id == assignment_id,
            Assignment.school_id == school_id,
        )
    )
    row = result.first()

    if not row:
        raise HTTPException(status_code=404, detail="Assignment not found")

    assignment, course_name = row

    return {
        "data": {
            "id": assignment.id,
            "title": assignment.title,
            "course_id": assignment.course_id,
            "course_name": course_name,
            "due_date": assignment.due_date.isoformat() if assignment.due_date else None,
            "status": assignment.status,
            "description": assignment.description,
            "instructions": assignment.instructions,
            "assignment_type": assignment.assignment_type,
            "total_marks": assignment.total_marks,
        }
    }


@router.delete("/{assignment_id}")
async def delete_assignment(assignment_id: int, db: AsyncSession = Depends(get_db), current_user=Depends(get_current_user)):
    user, payload = current_user
    school_id = payload.get("school_id")

    result = await db.execute(
        select(Assignment).where(
            Assignment.id == assignment_id,
            Assignment.school_id == school_id,
            Assignment.teacher_id == user.id,
        )
    )
    assignment = result.scalar_one_or_none()
    if not assignment:
        raise HTTPException(status_code=404, detail="Assignment not found")

    await db.delete(assignment)
    await db.commit()
    return {"message": "Assignment deleted."}


@router.get("/{assignment_id}/submissions")
async def get_assignment_submissions(assignment_id: int, db: AsyncSession = Depends(get_db), current_user=Depends(get_current_user)):
    """Fetch all submissions for an assignment (includes pending students who haven't submitted yet)."""
    result = await db.execute(
        select(AssignmentSubmission).where(AssignmentSubmission.assignment_id == assignment_id)
    )
    submissions = result.scalars().all()

    data = []
    for sub in submissions:
        student_result = await db.execute(
            select(Student).where(Student.id == sub.student_id)
        )
        student = student_result.scalar_one_or_none()

        data.append({
            "id": sub.id,
            "student_id": sub.student_id,
            "first_name": student.first_name if student else "Unknown",
            "last_name": student.last_name if student else "Student",
            "admission_number": student.admission_number if student else None,
            "status": sub.status,
            "score": sub.grade,
            "max_grade": sub.max_grade,
            "is_late": sub.is_late,
            "submitted_at": sub.submitted_at.isoformat() if sub.submitted_at else None,
            "submission_text": sub.submission_text,
            "feedback": sub.feedback,
        })

    return {"data": data}


@router.post("/submissions/{submission_id}/grade")
async def grade_submission(
    submission_id: int,
    payload: GradeSubmissionRequest,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user)
):
    """Grade a specific submission and leave feedback"""
    user, _ = current_user  # Only need the user ID here

    result = await db.execute(
        select(AssignmentSubmission).where(AssignmentSubmission.id == submission_id)
    )
    submission = result.scalar_one_or_none()

    if not submission:
        raise HTTPException(status_code=404, detail="Submission not found")

    submission.grade = payload.grade
    submission.feedback = payload.feedback
    submission.status = "graded"
    submission.graded_by = user.id
    submission.graded_at = datetime.utcnow()

    await db.commit()
    return {"message": "Submission graded successfully", "data": {"id": submission.id, "score": submission.grade}}
